### PR TITLE
HOSTEDCP-2035: Use Client Cert Auth for ARO HCP deployments

### PR DESCRIPTION
**Pull request by bryan-cox**
Use Client Certificate Authentication for ARO HCP deployments. HyperShift will pass the needed environment variables for this authentication method: ARO_HCP_MI_CLIENT_ID, ARO_HCP_TENANT_ID, and ARO_HCP_CLIENT_CERTIFICATE_PATH.

### CODE REVIEW

File: pkg/cloudprovider/azure.go [hunk 1]
```diff
@@ -586,17 +589,47 @@ func (a *Azure) getAuthorizer(env azureapi.Environment, cfg *azureCredentialsCon
...
+		if err != nil {
+			return nil, fmt.Errorf(`failed to read certificate file "%s": %v`, certPath, err)
+		}
+
+		certs, key, err := azidentity.ParseCertificates(certData, []byte{})
+		if err != nil {
+			return nil, fmt.Errorf(`failed to parse certificate data "%s": %v`, certPath, err)
+		}
+
+		// Set up a watch on our config file; if it changes, we should exit -
```
Comment (kyrtapz) [inline pkg/cloudprovider/azure.go:616]: config file? it is meant to exit if the cert changes right?
More broadly, is there a plan to add cert reload capability to the ARO go client?

File: pkg/cloudprovider/azure.go [hunk 2]
```diff
@@ -586,17 +589,47 @@ func (a *Azure) getAuthorizer(env azureapi.Environment, cfg *azureCredentialsCon
...
+		// (we don't have the ability to dynamically reload config changes).
+		stopCh := make(chan struct{})
+		err = interrupt.New(func(s os.Signal) {
+			_, _ = fmt.Fprintf(os.Stderr, "interrupt: Gracefully shutting down ...\n")
```
Comment (kyrtapz) [inline pkg/cloudprovider/azure.go:620]: What is the purpose of logging like that?

File: pkg/cloudprovider/azure.go [hunk 3]
```diff
@@ -586,17 +589,47 @@ func (a *Azure) getAuthorizer(env azureapi.Environment, cfg *azureCredentialsCon
...
+			close(stopCh)
+		}).Run(func() error {
+			if err := watchForChanges(certPath, stopCh); err != nil {
```
Comment (kyrtapz) [inline pkg/cloudprovider/azure.go:623]:
```suggestion
			return watchForChanges(certPath, stopCh)
```

File: pkg/cloudprovider/azure.go [hunk 4]
```diff
@@ -586,17 +589,47 @@ func (a *Azure) getAuthorizer(env azureapi.Environment, cfg *azureCredentialsCon
...
-		options := azidentity.ManagedIdentityCredentialOptions{
+	// Managed Identity Override for ARO HCP
+	managedIdentityClientID := os.Getenv("ARO_HCP_MI_CLIENT_ID")
+	if managedIdentityClientID != "" {
+		klog.Info("Using client certification Azure authentication for ARO HCP")
+		options := &azidentity.ClientCertificateCredentialOptions{
 			ClientOptions: azcore.ClientOptions{
 				Cloud: cloudConfig,
 			},
+			SendCertificateChain: true,
```
Comment (kyrtapz) [inline pkg/cloudprovider/azure.go:600]: Out of curiosity, could you comment on why is this required?

File: pkg/cloudprovider/azure.go [hunk 5]
```diff
@@ -658,3 +691,59 @@ func (a *Azure) getNodeLock(nodeName string) *sync.Mutex {
...
 	return id[strings.LastIndex(id, "/"):]
 }
+
+// watchForChanges closes stopCh if the configuration file changed.
+func watchForChanges(configPath string, stopCh chan struct{}) error {
+	configPath, err := filepath.Abs(configPath)
+	if err != nil {
+		return err
+	}
+	watcher, err := fsnotify.NewWatcher()
```
Comment (kyrtapz) [inline pkg/cloudprovider/azure.go:701]: Missing `defer watcher.Close()`

File: pkg/cloudprovider/azure.go [hunk 6]
```diff
@@ -658,3 +691,59 @@ func (a *Azure) getNodeLock(nodeName string) *sync.Mutex {
 func getNameFromResourceID(id string) string {
 	return id[strings.LastIndex(id, "/"):]
 }
+
+// watchForChanges closes stopCh if the configuration file changed.
```
Comment (kyrtapz) [inline pkg/cloudprovider/azure.go:695]: NIT: this function is not specific to config files, can you update the comment to better reflect what it does?

File: pkg/cloudprovider/azure.go [hunk 7]
```diff
@@ -658,3 +691,59 @@ func (a *Azure) getNodeLock(nodeName string) *sync.Mutex {
...
+				return err
+			}
+		} else {
+			break
+		}
+	}
+	go func() {
+		for {
+			select {
+			case <-stopCh:
```
Comment (kyrtapz) [inline pkg/cloudprovider/azure.go:730]: shouldn't this cause a return?

File: pkg/cloudprovider/azure.go [hunk 8]
```diff
@@ -658,3 +691,59 @@ func (a *Azure) getNodeLock(nodeName string) *sync.Mutex {
...
+				}
+				klog.V(2).Infof("Configuration file %s changed, exiting...", event.Name)
+				close(stopCh)
+				return
+			case err, ok := <-watcher.Errors:
+				if !ok {
+					klog.Errorf("failed to watch config file %s", p)
+					return
+				}
+				klog.Errorf("fsnotify error: %v", err)
```
Comment (kyrtapz) [inline pkg/cloudprovider/azure.go:744]: What will happen in this case? We will just stop watching for cert changes? shouldn't this close the channel?

File: pkg/cloudprovider/azure.go [hunk 9]
```diff
@@ -586,17 +589,47 @@ func (a *Azure) getAuthorizer(env azureapi.Environment, cfg *azureCredentialsCon
...
+		if err != nil {
+			return nil, fmt.Errorf(`failed to parse certificate data "%s": %v`, certPath, err)
+		}
+
+		// Set up a watch on our config file; if it changes, we should exit -
+		// (we don't have the ability to dynamically reload config changes).
+		stopCh := make(chan struct{})
+		err = interrupt.New(func(s os.Signal) {
+			_, _ = fmt.Fprintf(os.Stderr, "interrupt: Gracefully shutting down ...\n")
+			close(stopCh)
```
Comment (kyrtapz) [inline pkg/cloudprovider/azure.go:621]: I am confused about what actually happens when stopCh is closed and why is it closed here?

File: pkg/cloudprovider/azure.go (hunk 1 above)
Comment (bryan-cox) [reply pkg/cloudprovider/azure.go:616]: Yeah this is meant to exit (i.e. reboot the pod) when the cert changes. Azure is planning to add a reload capability to its SDK but it is not there yet.

File: pkg/cloudprovider/azure.go (hunk 2 above)
Comment (bryan-cox) [reply pkg/cloudprovider/azure.go:620]: I think this was just a copy/paste thing from somewhere. Happy to change this to klog.

File: pkg/cloudprovider/azure.go (hunk 4 above)
Comment (bryan-cox) [reply pkg/cloudprovider/azure.go:600]: This is related to 1st party Azure technology but with this, the certificate authentication doesn't require pinning certificates to applications, basically saying "Trust this certificate subject issued from this CA". cc @bennerv

File: pkg/cloudprovider/azure.go (hunk 6 above)
Comment (bryan-cox) [reply pkg/cloudprovider/azure.go:695]: Yeah I'll make it more generic

File: pkg/cloudprovider/azure.go (hunk 8 above)
Comment (bryan-cox) [reply pkg/cloudprovider/azure.go:744]: Yeah I believe so, we would stop watching for file changes. The channel should be closed on L621.

File: pkg/cloudprovider/azure.go (hunk 9 above)
Comment (bryan-cox) [reply pkg/cloudprovider/azure.go:621]: As I understand it, the caller of `watchForChanges` is ultimately responsible for determining if the channel needs to be closed or what needs done with it. So when `watchForChanges` returns it would pass the return back up the stack to `Run()` which passes it back to `New()`.

I copied this code from an example @enxebre put together, but it's essentially the same method used by OpenShift SDN [here](https://github.com/openshift/sdn/blob/18a57d7da14c7a45c4e8d55254bf7a89cee0ac43/pkg/cmd/openshift-sdn-node/cmd.go#L89).

File: pkg/cloudprovider/azure.go (hunk 7 above)
Comment (bryan-cox) [reply pkg/cloudprovider/azure.go:740]: No I don't believe so. I think we only want to return if there were changes or an error.

File: pkg/cloudprovider/azure.go (hunk 4 above)
Comment (bennerv) [reply pkg/cloudprovider/azure.go:600]: Note that it has no averse affect if you don't use SNI authentication with the 1p Azure technology.  The azure-sdk-for-go comment has a [bit more detail:](https://github.com/Azure/azure-sdk-for-go/blob/9591fab0ac0e5e3e2707d6db579331b136fb00ea/sdk/azidentity/client_certificate_credential.go#L45-L48) 
> 	// SendCertificateChain controls whether the credential sends the public certificate chain in the x5c
>	// header of each token request's JWT. This is required for Subject Name/Issuer (SNI) authentication.
>	// Defaults to False.

File: pkg/filewatcher/filewatcher.go [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+	go func() {
+		for {
+			select {
+			case <-done:
+				return
+			case <-ticker.C:
+				klog.Infof("Checking file for changes, %s", fileToWatch)
+				fileContents, err := os.ReadFile(fileToWatch)
+				if err != nil {
+					klog.Error("failed to read the file: %v", err)
```
Comment (kyrtapz) [inline pkg/filewatcher/filewatcher.go:51]: When a file read fails we will just stop watching it and if it changes this will never be caught.

File: pkg/filewatcher/filewatcher.go [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+		go checkForFileChanges(fileToWatch, interval)
+	})
+	return err
+}
+
+func checkForFileChanges(fileToWatch string, interval time.Duration) {
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+
+	done := make(chan bool)
```
Comment (kyrtapz) [inline pkg/filewatcher/filewatcher.go:40]: What's the point of this channel?

**Review by kyrtapz — CHANGES_REQUESTED** [review]
I am not confident in the new approach, It causes a hard stop and has a potentially significant delay.

File: pkg/filewatcher/filewatcher.go [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+
+	watchCertificateFileOnce.Do(func() {
+		klog.Infof("Starting the file change watcher on file, %s", fileToWatch)
+		fileContents, err = os.ReadFile(fileToWatch)
+		if err != nil {
+			err = fmt.Errorf("failed to read file on initialization: %v", err)
+			return
+		}
+
+		initialFileHash = hashSimple(fileContents)
```
[the remark this answers is not in the document]
Comment (bryan-cox) [reply pkg/filewatcher/filewatcher.go:30]: We can certainly poll sooner than 30m for the file check. How quick would you like it to poll?

File: pkg/filewatcher/filewatcher.go [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+	return err
+}
+
+func checkForFileChanges(fileToWatch string, interval time.Duration) {
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+
+	done := make(chan bool)
+
+	go func() {
```
[the remark this answers is not in the document]
Comment (bryan-cox) [reply pkg/filewatcher/filewatcher.go:42]: Yeah, I didn't need the second goroutine so I removed it.

File: pkg/filewatcher/filewatcher.go (hunk 10 above)
Comment (bryan-cox) [reply pkg/filewatcher/filewatcher.go:51]: The return should just return from the sync.Once function and the error would be returned at the end of `WatchFileForChanges`

File: pkg/filewatcher/filewatcher.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (bryan-cox) [reply pkg/filewatcher/filewatcher.go:38]: The reason for the periodic hash check - the previous file watcher code was not catching the changes when I was testing this function against the CPO. I could exec into the pod and see the file changed but the pod was not restarted nor did I see any messages about the file changing.

This current change seemed just as simple and required less code. This change was working when I was testing the same function against the CPO.

File: pkg/filewatcher/filewatcher.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (bryan-cox) [reply pkg/filewatcher/filewatcher.go:55]: Apparently this is needed. The pod does not restart when I removed this.
